rslib/src/scheduler/states/load_balancer.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::collections::HashSet;
+
+use rand::distributions::Distribution;
+use rand::distributions::WeightedIndex;
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+
+use super::fuzz::constrained_fuzz_bounds;
+use crate::card::CardId;
+use crate::decks::DeckId;
+use crate::notes::NoteId;
+use crate::prelude::Result;
+use crate::storage::SqliteStorage;
+
+const MAX_LOAD_BALANCE_INTERVAL: usize = 90;
+// due to the nature of load balancing, we may schedule things in the future and
+// so need to keep more than just the `MAX_LOAD_BALANCE_INTERVAL` days in our
+// cache. a flat 10% increase over the max interval should be enough to not have
+// problems
+const LOAD_BALANCE_DAYS: usize = (MAX_LOAD_BALANCE_INTERVAL as f32 * 1.1) as usize;
+const SIBLING_PENALTY: f32 = 0.001;
+
+#[derive(Debug, Default)]
+struct LoadBalancerDay {
+    cards: Vec<(CardId, NoteId)>,
+    notes: HashSet<NoteId>,
+}
+
+impl LoadBalancerDay {
+    fn add(&mut self, cid: CardId, nid: NoteId) {
+        self.cards.push((cid, nid));
+        self.notes.insert(nid);
+    }
+
+    fn remove(&mut self, cid: CardId) {
+        if let Some(index) = self.cards.iter().position(|c| c.0 == cid) {
+            let (_, rnid) = self.cards.swap_remove(index);
+
+            // if all cards of a note are removed, remove note
+            if !self.cards.iter().any(|(_cid, nid)| *nid == rnid) {
+                self.notes.remove(&rnid);
+            }
+        }
+    }
+
+    fn has_sibling(&self, nid: &NoteId) -> bool {
+        self.notes.contains(nid)
+    }
+}
+
+pub struct LoadBalancerContext<'a> {
+    load_balancer: &'a LoadBalancer,
+    note_id: Option<NoteId>,
+    fuzz_seed: Option<u64>,
+}
+
+impl<'a> LoadBalancerContext<'a> {
+    pub fn find_interval(&self, interval: f32, minimum: u32, maximum: u32) -> Option<u32> {
+        self.load_balancer
+            .find_interval(interval, minimum, maximum, self.fuzz_seed, self.note_id)
+    }
+
+    pub fn set_fuzz_seed(mut self, fuzz_seed: Option<u64>) -> Self {
+        self.fuzz_seed = fuzz_seed;
+        self
+    }
+}
+
+#[derive(Debug)]
+pub struct LoadBalancer {
+    days: [LoadBalancerDay; LOAD_BALANCE_DAYS],
+}
+
+impl LoadBalancer {
+    pub fn new(today: u32, deck_id: DeckId, storage: &SqliteStorage) -> Result<LoadBalancer> {
```

Comment:
If I'm following correctly, this is initialized with the deck ID of the deck the user clicked on, and you then filter out any due cards that aren't using the same preset. I don't think this is what the others intended when they were speaking about balancing per-preset. My assumption is that we'd partition the gathered cards into a map of config_id → cards, and then at balancing time, look up the count by the config_id of the current card.

If you weren't planning to do that for the MVP, perhaps it would make more sense to remove the filtering step, so balancing is global instead of simply disabled for presets other than the clicked-on deck.

Reply:
Are those two things not the same? I think I am not fully understanding something here.

Reply:
The deck_id at queue initialization is the one the user clicked on, and is only useful for settings that apply globally based on the current preset (e.g. the settings in the display order section). Contrast something like burying, where it depends not on the deck that was clicked on, but the preset that each card happens to use.

Reply:
Oh I am just now realizing that I forgot that deck heirarchies exist and multiple decks (thus presets) can be in the same review session.

Yeah this is a big oversight, should be easy to fix at least

Reply:
Consider the nested deck case, presets in brackets.

a(1)::b(2)

If the user clicks on `a`, and we are answering a card in `b`, we want to balance its interval based on number of cards in preset 2, not the cards in preset 1.

Reply:
(sorry, your post wasn't visible when I wrote the reply above)